lookout/core/metrics.py

```diff
@@ -88,9 +89,10 @@ def submit_event(key: str, value: Union[int, float, bool], description: str = ""
     :param description: Additional description of the event. Only used when creating a new event.
     :return: None
     """
-    global _prometheus_server
-    if _prometheus_server is None:
-        _prometheus_server = PrometheusServer(host=PROMETHEUS_HOST, port=PROMETHEUS_PORT)
+    global _prometheus_server, _prometheus_lock
+    with _prometheus_lock:
+        if _prometheus_server is None:
+            _prometheus_server = PrometheusServer(host=PROMETHEUS_HOST, port=PROMETHEUS_PORT)
     _prometheus_server.submit_event(key=key, value=value, description=description)
```

Comment:
Why can't we indent this line and remove the other two locks? (I'm taking this as a chance to learn a bit more about multiprocessing)

Reply:
Because the critical sections should be as short as possible. The global lock defends the server creation, then the metrics lock defends the counter creation, finally an individual counter lock defends the counter mutation. The parts of the code which can run in parallel run in parallel. That is, readonly ops and independent local calculations.

Ofc we could put everything under a single global lock and that would work correctly, however the performance would be much worse.